src/scripts/client/modules/hw.js

```diff
@@ -0,0 +1,200 @@
+   * event for a submission_id when clicked.  If a submission_id cannot be cloned
+   * display a revelant message to the user.
+   *
+   * @param {jQuery Element} $el the element to append the relevant link
+   * to
+   * @param {integer} submission_id the submission_id that is to be passed to
+   * NewlineHW
+   * @param {object} options configuration options for the link (currently
+   * on success class and fail class)
+   */
+  function addCloneLinkForSubmissionTo( $el, submission_id, options = {} ) {
+    isSubmissionCloneable( submission_id, function handleIsSubmissionCloneable( resp ) {
+      if ( resp.data.cloneable ) {
+        const clone_link = $el.append(
+          `
+          <a class="clone_and_open_submission ${options.success_class}">
+            <i class="fa fa-download" aria-hidden="true"></i>
+            Clone Assignment Locally
+          </a>
+          `
+        );
+
+        clone_link
+          .find( '.clone_and_open_submission' )
+          .click( function getSubmissionIDandCall( e ) {
+            e.preventDefault();
+            const self = $( this );
+
+            self.append( `
+              <i class='newline_hw_working fa fa-spin fa-spinner'></i>
+            ` );
+
+            triggerCloneEventForSubmissionID( submission_id, function() {
+              self.find( ".newline_hw_working" ).remove();
+            } );
+          } );
+      } else {
+        $el.append(
+          `
+          <a target="_blank" href="${resp.data.submission_info.url}" class="${options.fail_class}">
+            <i class="fa fa-external-link" aria-hidden="true"></i>
+            Go To
+          </a>
+          `
+        );
+      }
+    } );
+  }
+
+  function detectNativeClient( callback ) {
+    sendMessageToBackgroundWorker( {
+      event: "heartbeat",
+      data: {}
+    }, callback );
+  }
+
+  function isSubmissionCloneable( submission_id, callback ) {
+    sendMessageToBackgroundWorker( {
+      event: "check_if_cloneable",
+      data: {
+        id: submission_id
+      }
+    }, callback );
+  }
+
+  function triggerCloneEventForSubmissionID( submission_id, callback ) {
+    sendMessageToBackgroundWorker( {
+      event: "clone_and_open_submission",
+      data: {
+        id: submission_id
+      }
+    }, callback );
+  }
+
+  /**
+   * send json objects to the backgroup process for privledge esclation.
+   *
+   * @param  {object}   msg      event data to be sent to native client
+   * @param  {Function} callback optional callback with response for native client
+   */
+  function sendMessageToBackgroundWorker( msg, callback ) {
+    callback = ( typeof( callback ) === "function" && callback ) || function() {};
+    chrome.runtime.sendMessage( msg, function( resp ) {
+      console.log( "From Background page", resp );
+
+      // Handle native client error message / timeout
+      if ( resp && resp.status === "fail" ) {
+        console.error( resp.message );
+        tiy.showMessage( resp.message );
```

Comment:
I think we want to keep going as this is a utility function it's possible that the caller may want access to the failure state, regardless of the common handling for one. 

Though it was here previously no caller currently uses this behavior.